examples/deep_learning/plot_robust_training.py

```diff
@@ -0,0 +1,229 @@
+  del argv
+
+  # Hide any GPUs from TensorFlow. Otherwise TF might reserve memory and make
+  # it unavailable to JAX.
+  tf.config.experimental.set_visible_devices([], "GPU")
+  train_ds, ds_info = load_dataset("train", is_training=True,
+                                   batch_size=FLAGS.train_batch_size)
+  iter_per_epoch = ds_info.splits["train"].num_examples // FLAGS.train_batch_size
+  input_shape = (1,) + ds_info.features["image"].shape
+  num_classes = ds_info.features["label"].num_classes
+
+  net = CNN(num_classes)
+
+  @jax.jit
+  def accuracy(params, data):
+    inputs, labels = data
+    x = inputs.astype(jnp.float32) / 255.
+    logits = net.apply({"params": params}, x)
+    return jnp.mean(jnp.argmax(logits, axis=-1) == labels)
+
+  logistic_loss = jax.vmap(loss.multiclass_logistic_loss)
+
+  @jax.jit
+  def loss_fun(params, l2reg, data):
+    """Compute the loss of the network."""
+    inputs, labels = data
+    x = inputs.astype(jnp.float32) / 255.
+    logits = net.apply({"params": params}, x)
+    sqnorm = tree_util.tree_l2_norm(params, squared=True)
+    loss_value = jnp.mean(logistic_loss(labels, logits))
+    return loss_value + 0.5 * l2reg * sqnorm
+
+  @jax.jit
+  def pgd_attack(image, label, params, l2reg=0, epsilon=0.1, maxiter=10):
+    """PGD attack on the L-infinity ball with radius epsilon.
+
+    Args:
+      image: array-like, input data for the CNN
+      label: integer, class label corresponding to image
+      params: tree, parameters of the model to attack
+      l2reg: float, L2 regularization scale in the loss function
+      epsilon: float, radius of the L-infinity ball.
+      maxiter: int, number of iterations of this algorithm.
+
+    Returns:
+      perturbed_image: Adversarial image on the boundary of the L-infinity ball
+        of radius epsilon and centered at image.
+
+    Notes:
+      PGD attack is described in (Madry et al. 2017),
+      https://arxiv.org/pdf/1706.06083.pdf
+    """
+    image_perturbation = jnp.zeros_like(image)
+    def adversarial_loss(perturbation):
+      return loss_fun(params, l2reg, (image + perturbation, label))
+
+    for _ in range(maxiter):
+      # compute gradient of the loss wrt to the image
+      grad = jax.grad(adversarial_loss)(image_perturbation)
+
+      # heuristic step-size 2 eps / maxiter
+      image_perturbation += (2 * epsilon / maxiter) * jnp.sign(grad)
+      # projection step onto the L-infinity ball centered at image
+      image_perturbation = jnp.clip(image_perturbation, - epsilon, epsilon)
+
+    # clip the image to ensure pixels are between 0 and 1
+    return jnp.clip(image + image_perturbation, 0, 1)
+
+  # Initialize solver and parameters.
+  solver = OptaxSolver(
+      opt=optax.adam(FLAGS.learning_rate),
+      fun=loss_fun,
+      maxiter=FLAGS.epochs * iter_per_epoch)
+  rng = jax.random.PRNGKey(0)
+  params = net.init(rng, jnp.zeros(input_shape))["params"]
+
+  state = solver.init_state(params)
+  rng = jax.random.PRNGKey(0)
```

Comment:
Would it be better to split the rngs, rather than initialize the same one multiple times? This seems bug prone if someone reuses the code as a template.